server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToPlainText.java

```diff
@@ -0,0 +1,144 @@
+                mail.getMessage().saveChanges();
+            }
+        } catch (Exception e) {
+            throw new MessagingException("Exception while extracting HTML", e);
+        }
+    }
+
+    // true if the message content is mutated
+    private boolean removeHtmlFromMultipartAlternative(MimeMessage mimeMessage) throws IOException, MessagingException {
+        if (mimeMessage.getContent() instanceof Multipart) {
+            Multipart multipart = (Multipart) mimeMessage.getContent();
+            return removeHtmlFromMultipartAlternativeForContent(multipart);
+        }
+        return false;
+    }
+
+    // true if the message content is mutated
+    private boolean removeHtmlFromMultipartAlternative(BodyPart bodyPart) throws IOException, MessagingException {
+        if (bodyPart.getContent() instanceof Multipart) {
+            Multipart multipart = (Multipart) bodyPart.getContent();
+            return removeHtmlFromMultipartAlternativeForContent(multipart);
+        }
+        return false;
+    }
+
+    // true if the message content is mutated
+    private boolean removeHtmlFromMultipartAlternativeForContent(Multipart multipart) throws MessagingException, IOException {
+        boolean mutated = false;
+        if (multipart.getContentType().startsWith("multipart/alternative")) {
+            int removedParts = 0;
+            for (int i = 0; i < multipart.getCount(); i++) {
+                if (multipart.getBodyPart(i + removedParts).getContentType().startsWith("text/html")) {
+                    multipart.removeBodyPart(i + removedParts);
+                    removedParts++;
+                    mutated = true;
+                }
+            }
+        } else {
+            for (int i = 0; i < multipart.getCount(); i++) {
+                mutated = removeHtmlFromMultipartAlternative(multipart.getBodyPart(i));
+            }
+        }
+        return mutated;
+    }
+
+    // true if the message content is mutated
+    private boolean convertRemainingHtmlToPlainText(MimeMessage mimeMessage) throws IOException, MessagingException {
+        if (mimeMessage.getContentType().startsWith("text/html")) {
+            mimeMessage.setContent(htmlTextExtractor.toPlainText(IOUtils.toString(mimeMessage.getInputStream())), "text/plain");
+            return true;
+        }
+        if (mimeMessage.getContent() instanceof Multipart) {
+            boolean mutated = false;
+            Multipart multipart = (Multipart) mimeMessage.getContent();
+            for (int i = 0; i < multipart.getCount(); i++) {
+                mutated = convertRemainingHtmlToPlainText(multipart.getBodyPart(i));
+            }
+            return mutated;
+        }
+        return false;
+    }
+
+    // true if the message content is mutated
+    private boolean convertRemainingHtmlToPlainText(BodyPart bodyPart) throws IOException, MessagingException {
+        if (bodyPart.getContent() instanceof Multipart) {
```

Comment:
This if clause and the one in line 112 are exactly the same code inside. Extracting it into a private method maybe would be better?